#
# Upstream bug:
#   https://gcc.gnu.org/bugzilla/show_bug.cgi?id=79749
# Upstream changeset:
#   https://gcc.gnu.org/viewcvs?rev=245793&root=gcc&view=rev
#
# Index: gcc-5-branch/gcc/ChangeLog
# ===================================================================
# --- gcc-5-branch/gcc/ChangeLog	(revision 245792)
# +++ gcc-5-branch/gcc/ChangeLog	(revision 245793)
# @@ -1,3 +1,9 @@
# +2017-02-28  Eric Botcazou  <ebotcazou@adacore.com>
# +
# +	PR target/79749
# +	* config/sparc/sparc.c (sparc_frame_pointer_required): Add missing
# +	condition on optimize for the leaf function test.
# +
#  2017-02-22  Bill Schmidt  <wschmidt@linux.vnet.ibm.com>
#  
#  	Backport from mainline
Index: gcc-5-branch/gcc/testsuite/gcc.target/sparc/20170228-1.c
===================================================================
--- gcc-5-branch/gcc/testsuite/gcc.target/sparc/20170228-1.c	(nonexistent)
+++ gcc-5-branch/gcc/testsuite/gcc.target/sparc/20170228-1.c	(revision 245793)
@@ -0,0 +1,20 @@
+/* PR target/79749 */
+/* Reported by Rainer Orth <ro@gcc.gnu.org> */
+
+/* { dg-do run } */
+/* { dg-options "-fomit-frame-pointer" } */
+
+extern void abort (void);
+
+int foo (int x1, int x2, int x3, int x4, int x5, int x6, int x7)
+{
+  return x7;
+}
+
+int main (void)
+{
+  if (foo (100, 200, 300, 400, 500, 600, 700) != 700)
+    abort ();
+
+  return 0;
+}
# Index: gcc-5-branch/gcc/testsuite/ChangeLog
# ===================================================================
# --- gcc-5-branch/gcc/testsuite/ChangeLog	(revision 245792)
# +++ gcc-5-branch/gcc/testsuite/ChangeLog	(revision 245793)
# @@ -1,3 +1,7 @@
# +2017-02-28  Eric Botcazou  <ebotcazou@adacore.com>
# +
# +	* gcc.target/sparc/20170228-1.c: New test.
# +
#  2017-02-24  Eric Botcazou  <ebotcazou@adacore.com>
#  
#  	* gnat.dg/opt63.adb: New test.
Index: gcc-5-branch/gcc/config/sparc/sparc.c
===================================================================
--- gcc-5-branch/gcc/config/sparc/sparc.c	(revision 245792)
+++ gcc-5-branch/gcc/config/sparc/sparc.c	(revision 245793)
@@ -11989,8 +11989,9 @@
   if (TARGET_FLAT)
     return false;
 
-  /* Otherwise, the frame pointer is required if the function isn't leaf.  */
-  return !(crtl->is_leaf && only_leaf_regs_used ());
+  /* Otherwise, the frame pointer is required if the function isn't leaf, but
+     we cannot use sparc_leaf_function_p since it hasn't been computed yet.  */
+  return !(optimize > 0 && crtl->is_leaf && only_leaf_regs_used ());
 }
 
 /* The way this is structured, we can't eliminate SFP in favor of SP
